File: devcollab-worker/src/main/java/com/devcollab/worker/observability/RuntimeMemoryProfiler.java
package com.devcollab.worker.observability;

import com.fasterxml.jackson.databind.ObjectMapper;
import jakarta.annotation.PreDestroy;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

import java.io.BufferedWriter;
import java.io.IOException;
import java.lang.management.BufferPoolMXBean;
import java.lang.management.GarbageCollectorMXBean;
import java.lang.management.ManagementFactory;
import java.lang.management.MemoryPoolMXBean;
import java.lang.management.MemoryUsage;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardOpenOption;
import java.time.Instant;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicBoolean;
import java.util.concurrent.atomic.AtomicLong;
import java.util.regex.Pattern;

@Component
public class RuntimeMemoryProfiler implements AutoCloseable {

    private static final Logger LOGGER = LoggerFactory.getLogger(RuntimeMemoryProfiler.class);
    private static final Pattern SAFE_RUN_ID = Pattern.compile("[A-Za-z0-9][A-Za-z0-9._-]{0,63}");
    private static final int ATTRIBUTE_LIMIT = 24;
    private static final int ATTRIBUTE_STRING_LIMIT = 256;
    private static final int SCHEMA_VERSION = 1;
    private static final Entry STOP = new Entry(false, "");

    private final ObjectMapper objectMapper = new ObjectMapper();
    private final ArrayBlockingQueue<Entry> queue;
    private final AtomicBoolean active = new AtomicBoolean();
    private final AtomicLong samplesWritten = new AtomicLong();
    private final AtomicLong eventsWritten = new AtomicLong();
    private final AtomicLong droppedRecords = new AtomicLong();
    private final AtomicLong writeErrors = new AtomicLong();
    private final ThreadLocal<Stage> currentStage = new ThreadLocal<>();
    private final String runId;
    private final String service;
    private final String instanceId;
    private final long pid;
    private final long intervalMs;
    private final Path sampleFile;
    private final Path eventFile;
    private Thread samplerThread;
    private Thread writerThread;

    @Autowired
    public RuntimeMemoryProfiler(
            @Value("${devcollab.memory-profile.enabled:false}") boolean enabled,
            @Value("${devcollab.memory-profile.run-id:}") String configuredRunId,
            @Value("${devcollab.memory-profile.output-dir:.data/memory-profile}") String outputDir,
            @Value("${devcollab.memory-profile.interval-ms:1000}") long configuredIntervalMs,
            @Value("${devcollab.memory-profile.queue-capacity:1024}") int queueCapacity,
            @Value("${spring.application.name:devcollab-worker}") String service
    ) {
        this(enabled, configuredRunId, Path.of(outputDir), configuredIntervalMs, queueCapacity,
                service);
    }

    RuntimeMemoryProfiler(
            boolean enabled,
            String configuredRunId,
            Path outputDir,
            long configuredIntervalMs,
            int queueCapacity,
            String service
    ) {
        this.service = safeService(service);
        this.pid = ProcessHandle.current().pid();
        this.instanceId = this.service + "-" + pid + "-" + UUID.randomUUID().toString().substring(0, 8);
        this.intervalMs = configuredIntervalMs >= 500 ? configuredIntervalMs : 1000;
        this.queue = new ArrayBlockingQueue<>(Math.max(1, Math.min(queueCapacity, 65_536)));
        this.runId = enabled ? safeRunId(configuredRunId) : "disabled";
        Path runDirectory = outputDir.toAbsolutePath().normalize().resolve(this.runId).normalize();
        Path root = outputDir.toAbsolutePath().normalize();
        if (!runDirectory.startsWith(root)) {
            throw new IllegalArgumentException("Memory profile run directory escapes output root");
        }
        this.sampleFile = runDirectory.resolve(this.service + "-" + pid + "-samples.jsonl");
        this.eventFile = runDirectory.resolve(this.service + "-" + pid + "-events.jsonl");
        if (enabled) start(runDirectory);
    }

    public boolean enabled() {
        return active.get();
    }

    public Stage stage(
            String stage,
            String jobId,
            String repositoryId,
            String revision,
            String unitId
    ) {
        if (!active.get()) return Stage.noop();
        Stage parent = currentStage.get();
        Stage created = new Stage(
                this, safeStage(stage), jobId, repositoryId, revision, unitId, parent
        );
        currentStage.set(created);
        return created;
    }

    long droppedRecordCount() {
        return droppedRecords.get();
    }

    Map<String, Object> runtimeSample() {
        MemoryUsage heap = ManagementFactory.getMemoryMXBean().getHeapMemoryUsage();
        MemoryUsage nonHeap = ManagementFactory.getMemoryMXBean().getNonHeapMemoryUsage();
        BufferPoolMXBean direct = ManagementFactory.getPlatformMXBeans(BufferPoolMXBean.class)
                .stream().filter(pool -> "direct".equals(pool.getName())).findFirst().orElse(null);
        long gcCount = ManagementFactory.getGarbageCollectorMXBeans().stream()
                .mapToLong(GarbageCollectorMXBean::getCollectionCount)
                .filter(value -> value >= 0).sum();
        long gcTime = ManagementFactory.getGarbageCollectorMXBeans().stream()
                .mapToLong(GarbageCollectorMXBean::getCollectionTime)
                .filter(value -> value >= 0).sum();
        var threads = ManagementFactory.getThreadMXBean();
        var runtime = ManagementFactory.getRuntimeMXBean();
        var baseOs = ManagementFactory.getOperatingSystemMXBean();
        Long virtualMemory = null;
        Double cpuPercent = null;
        Long openFiles = null;
        if (baseOs instanceof com.sun.management.OperatingSystemMXBean os) {
            virtualMemory = nullableNonNegative(os.getCommittedVirtualMemorySize());
            double load = os.getProcessCpuLoad();
            cpuPercent = load < 0 ? null : load * 100.0;
        }
        if (baseOs instanceof com.sun.management.UnixOperatingSystemMXBean unix) {
            openFiles = nullableNonNegative(unix.getOpenFileDescriptorCount());
        }
        Long metaspace = ManagementFactory.getMemoryPoolMXBeans().stream()
                .filter(pool -> "Metaspace".equals(pool.getName()))
                .map(MemoryPoolMXBean::getUsage)
                .filter(usage -> usage != null)
                .map(MemoryUsage::getUsed)
                .findFirst().orElse(null);

        Map<String, Object> record = common("sample");
        record.put("rssBytes", linuxRssBytes());
        record.put("virtualMemoryBytes", virtualMemory);
        record.put("cpuPercent", cpuPercent);
        record.put("threadCount", threads.getThreadCount());
        record.put("peakThreadCount", threads.getPeakThreadCount());
        record.put("openFileDescriptorCount", openFiles);
        record.put("heapUsedBytes", heap.getUsed());
        record.put("heapCommittedBytes", heap.getCommitted());
        record.put("heapMaxBytes", heap.getMax());
        record.put("nonHeapUsedBytes", nonHeap.getUsed());
        record.put("metaspaceUsedBytes", metaspace);
        record.put("directBufferUsedBytes", direct == null ? null : direct.getMemoryUsed());
        record.put("directBufferCapacityBytes", direct == null ? null : direct.getTotalCapacity());
        record.put("directBufferCount", direct == null ? null : direct.getCount());
        record.put("gcCount", gcCount);
        record.put("gcTimeMs", gcTime);
        record.put("pythonAllocatedBytes", null);
        record.put("pythonPeakAllocatedBytes", null);
        record.put("processUptimeMs", runtime.getUptime());
        record.put("samplesWritten", samplesWritten.get());
        record.put("eventsWritten", eventsWritten.get());
        record.put("droppedRecords", droppedRecords.get());
        record.put("writeErrors", writeErrors.get());
        record.put("profilingOverheadWarning", droppedRecords.get() > 0);
        return record;
    }

    private void start(Path runDirectory) {
        try {
            Files.createDirectories(runDirectory);
            active.set(true);
            writerThread = new Thread(this::writerLoop, service + "-memory-profile-writer");
            samplerThread = new Thread(this::samplerLoop, service + "-memory-profile-sampler");
            writerThread.setDaemon(true);
            samplerThread.setDaemon(true);
            writerThread.start();
            samplerThread.start();
        } catch (Exception exception) {
            active.set(false);
            LOGGER.warn("Runtime memory profiling disabled because output initialization failed: {}",
                    exception.getClass().getSimpleName());
        }
    }

    private void samplerLoop() {
        while (active.get()) {
            enqueue(false, runtimeSample());
            try {
                Thread.sleep(intervalMs);
            } catch (InterruptedException exception) {
                Thread.currentThread().interrupt();
                return;
            } catch (RuntimeException exception) {
                LOGGER.warn("Runtime memory sampling disabled: {}",
                        exception.getClass().getSimpleName());
                active.set(false);
            }
        }
    }

    private void writerLoop() {
        try (BufferedWriter samples = writer(sampleFile);
             BufferedWriter events = writer(eventFile)) {
            while (active.get() || !queue.isEmpty()) {
                Entry entry = queue.poll(250, TimeUnit.MILLISECONDS);
                if (entry == null) continue;
                if (entry == STOP) break;
                BufferedWriter target = entry.event ? events : samples;
                target.write(entry.json);
                target.newLine();
                if (entry.event) eventsWritten.incrementAndGet();
                else samplesWritten.incrementAndGet();
            }
            samples.flush();
            events.flush();
        } catch (Exception exception) {
            if (!(exception instanceof InterruptedException)) {
                writeErrors.incrementAndGet();
                LOGGER.warn("Runtime memory profiling writer disabled: {}",
                        exception.getClass().getSimpleName());
            }
            active.set(false);
            queue.clear();
            Thread.currentThread().interrupt();
        }
    }

    private BufferedWriter writer(Path path) throws IOException {
        return Files.newBufferedWriter(path, StandardCharsets.UTF_8,
                StandardOpenOption.CREATE, StandardOpenOption.APPEND);
    }

    private void stageEvent(Stage stage, String eventType, long durationMs, Throwable error) {
        Map<String, Object> record = common("stage_event");
        record.put("eventType", eventType);
        record.put("stage", stage.name);
        record.put("stageExecutionId", stage.executionId);
        record.put("parentStageExecutionId", stage.parentExecutionId);
        record.put("jobId", stage.jobId);
        record.put("repositoryId", stage.repositoryId);
        record.put("revision", stage.revision);
        record.put("unitId", stage.unitId);
        record.put("durationMs", durationMs < 0 ? null : durationMs);
        record.put("status", eventType);
        record.put("errorType", error == null ? null : error.getClass().getSimpleName());
        record.put("attributes", Map.copyOf(stage.attributes));
        enqueue(true, record);
    }

    private Map<String, Object> common(String recordType) {
        Map<String, Object> record = new LinkedHashMap<>();
        record.put("schemaVersion", SCHEMA_VERSION);
        record.put("recordType", recordType);
        record.put("timestampUtc", Instant.now().toString());
        record.put("monotonicNanos", System.nanoTime());
        record.put("runId", runId);
        record.put("service", service);
        record.put("instanceId", instanceId);
        record.put("pid", pid);
        return record;
    }

    private void enqueue(boolean event, Map<String, Object> record) {
        if (!active.get()) return;
        try {
            String json = objectMapper.writeValueAsString(record);
            if (!queue.offer(new Entry(event, json))) droppedRecords.incrementAndGet();
        } catch (Exception exception) {
            droppedRecords.incrementAndGet();
        }
    }

    @PreDestroy
    @Override
    public void close() {
        if (!active.getAndSet(false)) return;
        if (samplerThread != null) samplerThread.interrupt();
        queue.offer(STOP);
        join(samplerThread);
        join(writerThread);
    }

    private void join(Thread thread) {
        if (thread == null || thread == Thread.currentThread()) return;
        try {
            thread.join(2_000);
        } catch (InterruptedException exception) {
            if (exception instanceof InterruptedException) {
                Thread.currentThread().interrupt();
            }
        }
    }

    private static String safeRunId(String value) {
        String candidate = value == null || value.isBlank()
                ? Instant.now().toString().replaceAll("[^0-9]", "").substring(0, 14)
                : value.trim();
        if (!SAFE_RUN_ID.matcher(candidate).matches()) {
            throw new IllegalArgumentException("Invalid memory profile runId");
        }
        return candidate;
    }

    private static String safeService(String value) {
        String candidate = value == null ? "service" : value.trim();
        String cleaned = candidate.replaceAll("[^A-Za-z0-9._-]", "-");
        if (cleaned.isBlank()) return "service";
        return cleaned.substring(0, Math.min(cleaned.length(), 64));
    }

    private static String safeStage(String value) {
        if (value == null || !value.matches("[A-Z][A-Z0-9_]{0,63}")) {
            throw new IllegalArgumentException("Invalid memory profile stage");
        }
        return value;
    }

    private static Long nullableNonNegative(long value) {
        return value < 0 ? null : value;
    }

    private static Long linuxRssBytes() {
        Path status = Path.of("/proc/self/status");
        if (!Files.isReadable(status)) return null;
        try {
            for (String line : Files.readAllLines(status, StandardCharsets.UTF_8)) {
                if (line.startsWith("VmRSS:")) {
                    String value = line.substring(6).trim().split("\\s+")[0];
                    return Long.parseLong(value) * 1024;
                }
            }
        } catch (Exception ignored) {
            return null;
        }
        return null;
    }

    private record Entry(boolean event, String json) {
    }

    public static final class Stage implements AutoCloseable {

        private static final Stage NOOP = new Stage();
        private final RuntimeMemoryProfiler profiler;
        private final String name;
        private final String executionId;
        private final String parentExecutionId;
        private final Stage parent;
        private final String jobId;
        private final String repositoryId;
        private final String revision;
        private final String unitId;
        private final long startedNanos;
        private final Map<String, Object> attributes = new LinkedHashMap<>();
        private Throwable failure;
        private boolean closed;

        private Stage() {
            profiler = null;
            name = "";
            executionId = "";
            parentExecutionId = null;
            parent = null;
            jobId = null;
            repositoryId = null;
            revision = null;
            unitId = null;
            startedNanos = 0;
            closed = true;
        }

        private Stage(
                RuntimeMemoryProfiler profiler,
                String name,
                String jobId,
                String repositoryId,
                String revision,
                String unitId,
                Stage parent
        ) {
            this.profiler = profiler;
            this.name = name;
            this.executionId = UUID.randomUUID().toString();
            this.parentExecutionId = parent == null ? null : parent.executionId;
            this.parent = parent;
            this.jobId = jobId;
            this.repositoryId = repositoryId;
            this.revision = revision;
            this.unitId = unitId;
            this.startedNanos = System.nanoTime();
            profiler.stageEvent(this, "STARTED", -1, null);
        }

        static Stage noop() {
            return NOOP;
        }

        public Stage attribute(String key, Object value) {
            if (profiler == null) return this;
            if (attributes.size() >= ATTRIBUTE_LIMIT && !attributes.containsKey(key)) {
                throw new IllegalArgumentException("Memory profile attribute limit exceeded");
            }
            if (key == null || !key.matches("[A-Za-z][A-Za-z0-9_]{0,63}")) {
                throw new IllegalArgumentException("Invalid memory profile attribute key");
            }
            if (!(value == null || value instanceof Number || value instanceof Boolean
                    || value instanceof String)) {
                throw new IllegalArgumentException("Memory profile attributes must be scalar");
            }
            if (value instanceof String text && text.length() > ATTRIBUTE_STRING_LIMIT) {
                throw new IllegalArgumentException("Memory profile attribute is too large");
            }
            attributes.put(key, value);
            return this;
        }

        public void failed(Throwable error) {
            this.failure = error;
        }

        @Override
        public void close() {
            if (closed || profiler == null) return;
            closed = true;
            long durationMs = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startedNanos);
            profiler.stageEvent(this, failure == null ? "COMPLETED" : "FAILED",
                    durationMs, failure);
            if (parentExecutionId == null) {
                profiler.currentStage.remove();
            } else {
                profiler.currentStage.set(findParent());
            }
        }

        private Stage findParent() {
            Stage current = profiler.currentStage.get();
            if (current != this) return null;
            return parentExecutionId == null ? null : parent;
        }
    }
}

File: agent-service/app/profiling/runtime_memory_profiler.py
from __future__ import annotations

import atexit
import contextvars
import json
import logging
import os
import queue
import re
import threading
import time
import tracemalloc
from collections.abc import Iterator
from contextlib import contextmanager
from dataclasses import dataclass, field
from datetime import UTC, datetime
from pathlib import Path
from typing import Any
from uuid import uuid4

import psutil  # type: ignore[import-untyped]

LOGGER = logging.getLogger("devcollab.memory_profile")
SCHEMA_VERSION = 1
SAFE_RUN_ID = re.compile(r"^[A-Za-z0-9][A-Za-z0-9._-]{0,63}$")
SAFE_STAGE = re.compile(r"^[A-Z][A-Z0-9_]{0,63}$")
SAFE_ATTRIBUTE = re.compile(r"^[A-Za-z][A-Za-z0-9_]{0,63}$")
MAX_ATTRIBUTES = 24
MAX_ATTRIBUTE_STRING = 256
_PARENT_STAGE: contextvars.ContextVar[str | None] = contextvars.ContextVar(
    "memory_profile_parent_stage", default=None
)


@dataclass(frozen=True)
class MemoryProfileConfig:
    enabled: bool = False
    run_id: str = ""
    output_dir: Path = Path(".data/memory-profile")
    interval_ms: int = 1000
    queue_capacity: int = 1024

    def validated(self) -> MemoryProfileConfig:
        if not self.enabled:
            return self
        run_id = self.run_id.strip() or datetime.now(UTC).strftime("%Y%m%dT%H%M%SZ")
        if not SAFE_RUN_ID.fullmatch(run_id):
            raise ValueError("Invalid memory profile runId")
        interval = self.interval_ms if self.interval_ms >= 500 else 1000
        capacity = min(max(self.queue_capacity, 1), 65_536)
        return MemoryProfileConfig(True, run_id, self.output_dir, interval, capacity)


class RuntimeMemoryProfiler:
    def __init__(self, config: MemoryProfileConfig, service: str) -> None:
        self._config = config.validated()
        self._service = _safe_service(service)
        self._pid = os.getpid()
        self._instance_id = f"{self._service}-{self._pid}-{uuid4().hex[:8]}"
        self._queue: queue.Queue[tuple[str, dict[str, Any]] | None] = queue.Queue(
            maxsize=self._config.queue_capacity
        )
        self._active = threading.Event()
        self._stop = threading.Event()
        self._writer_failed = threading.Event()
        self._samples_written = 0
        self._events_written = 0
        self._dropped_records = 0
        self._write_errors = 0
        self._lock = threading.Lock()
        self._sampler: threading.Thread | None = None
        self._writer: threading.Thread | None = None
        self._process: psutil.Process | None = None
        self._sample_file: Path | None = None
        self._event_file: Path | None = None
        if self._config.enabled:
            self._start()
        atexit.register(self.close)

    @property
    def enabled(self) -> bool:
        return self._active.is_set()

    @property
    def stats(self) -> dict[str, int]:
        with self._lock:
            return {
                "samplesWritten": self._samples_written,
                "eventsWritten": self._events_written,
                "droppedRecords": self._dropped_records,
                "writeErrors": self._write_errors,
            }

    @contextmanager
    def stage(
        self,
        name: str,
        *,
        job_id: str | None = None,
        repository_id: str | None = None,
        revision: str | None = None,
        unit_id: str | None = None,
    ) -> Iterator[ProfileStage]:
        if not self.enabled:
            yield ProfileStage.noop()
            return
        if not SAFE_STAGE.fullmatch(name):
            raise ValueError("Invalid memory profile stage")
        stage = ProfileStage(
            profiler=self,
            name=name,
            execution_id=str(uuid4()),
            parent_execution_id=_PARENT_STAGE.get(),
            job_id=job_id,
            repository_id=repository_id,
            revision=revision,
            unit_id=unit_id,
        )
        token = _PARENT_STAGE.set(stage.execution_id)
        self._stage_event(stage, "STARTED")
        try:
            yield stage
        except BaseException as exc:
            self._stage_event(stage, "FAILED", exc)
            raise
        else:
            self._stage_event(stage, "COMPLETED")
        finally:
            _PARENT_STAGE.reset(token)

    def sample(self) -> dict[str, Any]:
        record = self._common("sample")
        rss: int | None = None
        vms: int | None = None
        cpu: float | None = None
        threads: int | None = None
        open_files: int | None = None
        uptime: int | None = None
        process = self._process
        if process is not None:
            try:
                memory = process.memory_info()
                rss, vms = memory.rss, memory.vms
                cpu = process.cpu_percent(interval=None)
                threads = process.num_threads()
                open_files = len(process.open_files())
                uptime = max(0, int((time.time() - process.create_time()) * 1000))
            except (psutil.Error, OSError):
                pass
        allocated: int | None = None
        peak: int | None = None
        if tracemalloc.is_tracing():
            allocated, peak = tracemalloc.get_traced_memory()
        record.update(
            rssBytes=rss,
            virtualMemoryBytes=vms,
            cpuPercent=cpu,
            threadCount=threads,
            openFileDescriptorCount=open_files,
            heapUsedBytes=None,
            heapCommittedBytes=None,
            heapMaxBytes=None,
            nonHeapUsedBytes=None,
            metaspaceUsedBytes=None,
            directBufferUsedBytes=None,
            directBufferCapacityBytes=None,
            directBufferCount=None,
            gcCount=None,
            gcTimeMs=None,
            pythonAllocatedBytes=allocated,
            pythonPeakAllocatedBytes=peak,
            processUptimeMs=uptime,
            **self.stats,
            profilingOverheadWarning=self._dropped_records > 0,
        )
        return record

    def close(self) -> None:
        if not self._active.is_set() and self._sampler is None:
            return
        self._active.clear()
        self._stop.set()
        if self._sampler is not None:
            self._sampler.join(timeout=2)
        try:
            self._queue.put_nowait(None)
        except queue.Full:
            with self._lock:
                self._dropped_records += 1
        if self._writer is not None:
            self._writer.join(timeout=2)
        self._sampler = None
        self._writer = None
        if tracemalloc.is_tracing():
            tracemalloc.stop()

    def _start(self) -> None:
        run_root = self._config.output_dir.expanduser().resolve()
        run_dir = (run_root / self._config.run_id).resolve()
        if run_root not in run_dir.parents:
            raise ValueError("Memory profile run directory escapes output root")
        try:
            run_dir.mkdir(parents=True, exist_ok=True)
            self._sample_file = run_dir / f"{self._service}-{self._pid}-samples.jsonl"
            self._event_file = run_dir / f"{self._service}-{self._pid}-events.jsonl"
            self._process = psutil.Process(self._pid)
            self._process.cpu_percent(interval=None)
            tracemalloc.start(1)
            self._active.set()
            self._writer = threading.Thread(
                target=self._writer_loop,
                name=f"{self._service}-memory-profile-writer",
                daemon=True,
            )
            self._sampler = threading.Thread(
                target=self._sampler_loop,
                name=f"{self._service}-memory-profile-sampler",
                daemon=True,
            )
            self._writer.start()
            self._sampler.start()
        except (OSError, psutil.Error) as exc:
            self._active.clear()
            LOGGER.warning(
                "Runtime memory profiling disabled because output initialization failed: %s",
                type(exc).__name__,
            )

    def _sampler_loop(self) -> None:
        while self._active.is_set() and not self._stop.is_set():
            try:
                self._enqueue("sample", self.sample())
            except Exception as exc:
                LOGGER.warning("Runtime memory sampling disabled: %s", type(exc).__name__)
                self._active.clear()
                return
            self._stop.wait(self._config.interval_ms / 1000)

    def _writer_loop(self) -> None:
        assert self._sample_file is not None and self._event_file is not None
        try:
            with (
                self._sample_file.open("a", encoding="utf-8", buffering=64 * 1024) as samples,
                self._event_file.open("a", encoding="utf-8", buffering=64 * 1024) as events,
            ):
                while self._active.is_set() or not self._queue.empty():
                    try:
                        item = self._queue.get(timeout=0.25)
                    except queue.Empty:
                        continue
                    if item is None:
                        break
                    kind, record = item
                    target = events if kind == "stage_event" else samples
                    target.write(json.dumps(record, ensure_ascii=False, separators=(",", ":")))
                    target.write("\n")
                    with self._lock:
                        if kind == "stage_event":
                            self._events_written += 1
                        else:
                            self._samples_written += 1
                samples.flush()
                events.flush()
        except (OSError, TypeError, ValueError) as exc:
            with self._lock:
                self._write_errors += 1
            self._writer_failed.set()
            self._active.clear()
            _drain(self._queue)
            LOGGER.warning(
                "Runtime memory profiling writer disabled: %s", type(exc).__name__
            )

    def _enqueue(self, kind: str, record: dict[str, Any]) -> None:
        if not self.enabled or self._writer_failed.is_set():
            return
        try:
            self._queue.put_nowait((kind, record))
        except queue.Full:
            with self._lock:
                self._dropped_records += 1

    def _common(self, record_type: str) -> dict[str, Any]:
        return {
            "schemaVersion": SCHEMA_VERSION,
            "recordType": record_type,
            "timestampUtc": datetime.now(UTC).isoformat().replace("+00:00", "Z"),
            "monotonicNanos": time.monotonic_ns(),
            "runId": self._config.run_id,
            "service": self._service,
            "instanceId": self._instance_id,
            "pid": self._pid,
        }

    def _stage_event(
        self,
        stage: ProfileStage,
        event_type: str,
        error: BaseException | None = None,
    ) -> None:
        record = self._common("stage_event")
        duration = (
            None
            if event_type == "STARTED"
            else max(0, (time.monotonic_ns() - stage.started_nanos) // 1_000_000)
        )
        record.update(
            eventType=event_type,
            stage=stage.name,
            stageExecutionId=stage.execution_id,
            parentStageExecutionId=stage.parent_execution_id,
            jobId=stage.job_id,
            repositoryId=stage.repository_id,
            revision=stage.revision,
            unitId=stage.unit_id,
            durationMs=duration,
            status=event_type,
            errorType=type(error).__name__ if error is not None else None,
            attributes=dict(stage.attributes),
        )
        self._enqueue("stage_event", record)


@dataclass
class ProfileStage:
    profiler: RuntimeMemoryProfiler | None
    name: str
    execution_id: str
    parent_execution_id: str | None
    job_id: str | None
    repository_id: str | None
    revision: str | None
    unit_id: str | None
    started_nanos: int = field(default_factory=time.monotonic_ns)
    attributes: dict[str, str | int | float | bool | None] = field(default_factory=dict)

    @classmethod
    def noop(cls) -> ProfileStage:
        return cls(None, "", "", None, None, None, None, None)

    def attribute(self, key: str, value: object) -> ProfileStage:
        if self.profiler is None:
            return self
        if not SAFE_ATTRIBUTE.fullmatch(key):
            raise ValueError("Invalid memory profile attribute key")
        if len(self.attributes) >= MAX_ATTRIBUTES and key not in self.attributes:
            raise ValueError("Memory profile attribute limit exceeded")
        if value is not None and not isinstance(value, str | int | float | bool):
            raise ValueError("Memory profile attributes must be scalar")
        if isinstance(value, str) and len(value) > MAX_ATTRIBUTE_STRING:
            raise ValueError("Memory profile attribute is too large")
        self.attributes[key] = value
        return self


def _safe_service(value: str) -> str:
    cleaned = re.sub(r"[^A-Za-z0-9._-]", "-", value.strip())[:64]
    return cleaned or "service"


def _drain(items: queue.Queue[Any]) -> None:
    while True:
        try:
            items.get_nowait()
        except queue.Empty:
            return
